Main.hs

```diff
@@ -3,35 +3,88 @@ module Main where
 import Control.Monad (unless)
 import Text.Printf (printf)
 
-data Expr
+data Expr 
+  = Num Double
+  | Sqrt Expr 
+  | Add Expr Expr       
+  | Sub Expr Expr       
+  | Mul Expr Expr       
+  | Div Expr Expr       
+  | Pow Expr Expr                 
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = printf "%s" (show a)
+  show (Sqrt a) = printf ("Square root of %s") (show a) 
+  show (Add a b) = printf ("%s + %s") (show a) (show b)
+  show (Sub a b) = printf ("%s - %s") (show a) (show b)
+  show (Mul a b) = printf ("%s * %s") (show a) (show b)
+  show (Div a b) = printf ("%s / %s") (show a) (show b)
+  show (Pow a b) = printf ("%s ^ %s") (show a) (show b)
 
 instance Eq Expr where
-  (==) = undefined
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  _ == _ = False 
 
-data Error
+data Error = DivByZero | NegativeSqrt
```

Comment:
It would be grand to be able to see evaluation of which sub-expression caused the error.